### PR TITLE
PG-1382 PG-1383: sleepio community authentication

### PR DESCRIPTION
**Pull request by wsbastian**
**Reminder to create the PR against this repository, and not the base fork of `rafalp/Misago`.**

### JIRA Ticket

(subtask-Avi) https://bighealth.atlassian.net/browse/PG-1382
(subtask-Will) https://bighealth.atlassian.net/browse/PG-1383

### Problem

This is a shared feature branch, consolidating co-dependent work from two tickets listed above. This is in support of allowing users logged into Sleepio the ability to log into, and if needed, create accounts in the Community web-forum without presenting additional credentials.

### Solution

We have provided this experience through two mechanisms:
1. Defining custom middleware that exists in the Django middleware stack after session and user creation. This allows us to use existing mechanisms to log the user out/end the session if we are unable to refresh tokens. This middleware procures a Platform user_id (db key), and attaches it to the request which allows the social auth pipeline (below) to pull user information to create a user account
2. Misago uses https://python-social-auth.readthedocs.io/en/latest/ to drive the authentication and user creation from 3rd party sources. We have defined steps in this to source user IDs and create accounts based on the Platform user ID obtained in in the custom middleware

#### Unit tests + Integration Tests

Unit Tests: tests/community_app/test_middleware
Integration Tests: N/A

#### Manual Tests

- opening sleepio and community in a hosted environment in separate windows (same browser/not incognito. we need cookies to be shared)
- logging into sleepio
- accessing the view /complete/sleepio in community, and confirming that a user account is created (accepting yes to the terms of service)
- deleting the bh_access_token from the browser developer tools, doing any action in community (refresh, click link, etc) and ensuring the bh_refresh_token and bh_access_token are regenerated
- ensuring that these regenerated tokens allow users to continue to operate in sleepio
- logging out of sleepio, and ensuring this results in the user being logged out of community (this is one direction, logging out of community will not log out of sleepio, and logged out of community will be removed in M6)
- logging back into sleepio and again hitting /complete/sleepio in community, and ensuring your user is logged back in.

#### Related PRs

- https://github.com/sleepio/client-gateway-service-cluster/pull/52
- https://github.com/sleepio/client-gateway-service-cluster/pull/55

### CODE REVIEW

File: webapp/src/community_app/auth/backend.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+class SleepioAuth(BaseAuth):
+    name = "sleepio"
+    supports_inactive_user = False
+
+    def get_user_details(self, user_account):
+        return {
+            "username": user_account.get("uuid")[:12],  # TODO temporary
+            "email": user_account.get("email_address"),
+            "fullname": f"{user_account.get('first_name')} {user_account.get('last_name')}",
```
Comment (avi-bh) [inline webapp/src/community_app/auth/backend.py:19]: BP mentioned something about not showing this information. Do we want to also not store it (i.e. give it something meaningless)?

File: webapp/src/misago/socialauth/urls.py [hunk 2]
```diff
@@ -4,6 +4,6 @@
 
 
 urlpatterns = [
-    url(r"^login/(?P<backend>[^/]+)/$", views.auth, name="social-begin"),
-    url(r"^complete/(?P<backend>[^/]+)/$", views.complete, name="social-complete"),
+    # url(r"^login/(?P<backend>[^/]+)/$", views.auth, name="social-begin"),
```
Comment (avi-bh) [inline webapp/src/misago/socialauth/urls.py:7]: **nothing actionable for now:** I wish there was a better way of disabling these

**Review by avi-bh — APPROVED** [review]

File: webapp/src/community_app/auth/backend.py (hunk 1 above)
Comment (wsbastian) [reply webapp/src/community_app/auth/backend.py:19]: Good call, removed PII and inserted generic placeholder for email as it's required in account creation.

File: webapp/src/community_app/auth/backend.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+class SleepioAuth(BaseAuth):
+    name = "sleepio"
+    supports_inactive_user = False
+
+    def get_user_details(self, user_account):
+        return {
+            "username": user_account.get("uuid")[:12],  # TODO temporary
+            "uuid": user_account.get("uuid"),
+            # placeholder/fake address for all users
+            "email": "community_user@bighealth.com"
```
Comment (avi-bh) [inline webapp/src/community_app/auth/backend.py:20]: we actually might need real email address though. I believe it is required for sending notification emails for Community (like post responses etc)

Comment (avi-bh) [reply webapp/src/community_app/auth/backend.py:20]: worth checking with Venkat probably

Comment (wsbastian) [reply webapp/src/community_app/auth/backend.py:20]: mmm, I'd thought we wanted --no-- PII, but good point, i'll sync with BP and Venkat.

Comment (wsbastian) [reply webapp/src/community_app/auth/backend.py]: and emails are back in.

**Review by avi-bh — APPROVED** [review]

**Review by avi-bh — APPROVED** [review]

**Review by avi-bh — APPROVED** [review]

**Review by eliot-mestre — APPROVED** [review]

**Pull request merged**